File: gen-1/hermitcrab/archive/lib/bsp.py
# bsp.py — Block · Spindle · Point
# Semantic address resolver for pscale JSON blocks.
# Not code. Coordinates.

from typing import Any, Optional, Union
from dataclasses import dataclass, field

# --- Block loader registry ---
_block_loader = None


def bsp_register(loader):
    """Register a block loader so bsp can resolve block names."""
    global _block_loader
    _block_loader = loader


# --- Types ---

@dataclass
class SpindleNode:
    pscale: int      # depth below decimal: -1, -2, -3...
    digit: str       # the key at this level
    text: str        # content at this level (_ summary or leaf string)


@dataclass
class BspResult:
    mode: str                           # 'block' | 'spindle' | 'point'
    tree: Optional[dict] = None         # full tree (block mode)
    nodes: list = field(default_factory=list)  # spindle chain (spindle mode)
    text: Optional[str] = None          # focused content (point mode)


def bsp(
    block: Union[str, dict],
    spindle: Optional[float] = None,
    point: Optional[int] = None,
) -> BspResult:
    """
    bsp — Block · Spindle · Point

    bsp("wake")              → the full block tree
    bsp("wake", 0.842)       → spindle: nodes at 0.8, 0.84, 0.842
    bsp("wake", 0.842, 2)    → point: just the content at the final digit (2)

    block   - block name (string) resolved via registered loader, or a dict
    spindle - semantic number: each digit after the decimal is a tree key
    point   - focus digit within the spindle. Returns only that node's content.
    """
    # Resolve block name to object
    if isinstance(block, str):
        if _block_loader is None:
            return BspResult(mode="block", tree={})
        blk = _block_loader(block)
        if blk is None:
            return BspResult(mode="block", tree={})
    else:
        blk = block

    tree = blk.get("tree", blk)

    # Block mode — no spindle, return full tree
    if spindle is None:
        return BspResult(mode="block", tree=tree)

    # Parse semantic number into digit sequence
    s = f"{spindle:.10f}"
    dot = s.index(".")
    digits = s[dot + 1:].rstrip("0")

    # Walk the tree, building the spindle chain
    nodes = []
    node = tree

    for i, d in enumerate(digits):
        if not isinstance(node, dict) or d not in node:
            break
        node = node[d]
        if isinstance(node, str):
            text = node
        elif isinstance(node, dict) and "_" in node:
            text = node["_"]
        else:
            text = str(node)
        nodes.append(SpindleNode(pscale=-(i + 1), digit=d, text=text))

    if not nodes:
        return BspResult(mode="spindle", nodes=[])

    # Point mode — return just the focused node
    if point is not None:
        target = next((n for n in nodes if n.digit == str(point)), None)
        if target:
            return BspResult(mode="point", text=target.text)
        # Default: last node
        return BspResult(mode="point", text=nodes[-1].text)

    # Spindle mode — return the full chain
    return BspResult(mode="spindle", nodes=nodes)

File: gen-1/hermitcrab/archive/lib/bsp.ts
// bsp.ts — Block · Spindle · Point
// Semantic address resolver for pscale JSON blocks.
// Not code. Coordinates.

export interface PscaleBlock {
  decimal: number
  tree: Record<string, any>
}

export interface SpindleNode {
  pscale: number    // depth below decimal: -1, -2, -3...
  digit: string     // the key at this level
  text: string      // the content at this level (_ summary or leaf string)
}

export type BspResult =
  | { mode: 'block'; tree: Record<string, any> }
  | { mode: 'spindle'; nodes: SpindleNode[] }
  | { mode: 'point'; text: string }

/**
 * bsp — Block · Spindle · Point
 *
 * bsp("wake")              → the full block tree
 * bsp("wake", 0.842)       → spindle: nodes at 0.8, 0.84, 0.842
 * bsp("wake", 0.842, 2)    → point: just the content at the final digit (2)
 *
 * @param block  - block name (string) resolved via blockLoad, or a block object
 * @param spindle - semantic number: each digit after the decimal is a tree key
 * @param point   - focus digit within the spindle. If provided, returns only
 *                  that node's content. Omit for full spindle.
 */
export function bsp(
  block: string | PscaleBlock,
  spindle?: number,
  point?: number
): BspResult {
  // Resolve block name to object (caller provides blockLoad or passes object)
  const blk = typeof block === 'string'
    ? (globalThis as any).__bspBlockLoad?.(block) as PscaleBlock | undefined
    : block

  if (!blk?.tree) {
    return { mode: 'block', tree: {} }
  }

  // Block mode — no spindle, return full tree
  if (spindle === undefined) {
    return { mode: 'block', tree: blk.tree }
  }

  // Parse semantic number into digit sequence
  const str = spindle.toFixed(10)
  const dot = str.indexOf('.')
  const digits = dot === -1 ? [] : str.slice(dot + 1).replace(/0+$/, '').split('')

  // Walk the tree, building the spindle chain
  const nodes: SpindleNode[] = []
  let node: any = blk.tree

  for (let i = 0; i < digits.length; i++) {
    const d = digits[i]
    if (node[d] === undefined) break
    node = node[d]
    const text = typeof node === 'string'
      ? node
      : (typeof node === 'object' && node !== null && typeof node['_'] === 'string')
        ? node['_']
        : JSON.stringify(node)
    nodes.push({ pscale: -(i + 1), digit: d, text })
  }

  if (nodes.length === 0) {
    return { mode: 'spindle', nodes: [] }
  }

  // Point mode — return just the focused node
  if (point !== undefined) {
    // Find the node matching the requested digit
    const target = nodes.find(n => n.digit === String(point))
    if (target) {
      return { mode: 'point', text: target.text }
    }
    // Default: last node in spindle
    return { mode: 'point', text: nodes[nodes.length - 1].text }
  }

  // Spindle mode — return the full chain
  return { mode: 'spindle', nodes }
}

/**
 * Register a block loader so bsp can resolve block names.
 * Call once at kernel init: bspRegister(blockLoad)
 */
export function bspRegister(loader: (name: string) => PscaleBlock | undefined) {
  (globalThis as any).__bspBlockLoad = loader
}
